src/node_report.cc

```diff
@@ -0,0 +1,1115 @@
+
+#if !defined(_MSC_VER)
+#include <strings.h>
+#endif
+
+#ifdef _WIN32
+#include <Windows.h>
+#include <process.h>
+#include <dbghelp.h>
+#include <Lm.h>
+#include <tchar.h>
+#include <psapi.h>
+#else
+#include <sys/resource.h>
+// Get the standard printf format macros for C99 stdint types.
+#ifndef __STDC_FORMAT_MACROS
+#define __STDC_FORMAT_MACROS
+#endif
+#include <inttypes.h>
+#include <cxxabi.h>
+#include <dlfcn.h>
+#if defined(__linux__) || defined(__sun)
+#include <link.h>
+#endif
+#ifdef _AIX
+#include <sys/ldr.h>  // ld_info structure
+#endif
+// Include execinfo.h for the native backtrace API. The API is
+// unavailable on AIX and on some Linux distributions, e.g. Alpine Linux.
+#if !defined(_AIX) && !(defined(__linux__) && !defined(__GLIBC__))
+#include <execinfo.h>
+#endif
+#include <sys/utsname.h>
+#endif
+
+#ifdef __APPLE__
+#include <mach-o/dyld.h>  // _dyld_get_image_name()
+#endif
+
+
+
+#ifndef _WIN32
+extern char** environ;
+#endif
+
+namespace nodereport {
+
+using v8::Exception;
+using v8::HeapSpaceStatistics;
+using v8::HeapStatistics;
+using v8::Isolate;
+using v8::Local;
+using v8::Message;
+using v8::RegisterState;
+using v8::SampleInfo;
+using v8::String;
+using v8::V8;
+
+// Internal/static function declarations
+static void WriteNodeReport(Isolate* isolate, DumpEvent event,
+                            const char* message, const char* location,
+                            char* filename, std::ostream &out,
+                            MaybeLocal<Value> error, TIME_TYPE* time);
+static void PrintCommandLine(std::ostream& out);
+static void PrintVersionInformation(std::ostream& out);
+static void PrintJavaScriptStack(std::ostream& out, Isolate* isolate,
+                                 DumpEvent event, const char* location);
+static void PrintJavaScriptErrorStack(std::ostream& out, Isolate* isolate,
+                                      MaybeLocal<Value> error);
+static void PrintStackFromStackTrace(std::ostream& out, Isolate* isolate,
+                                     DumpEvent event);
+static void PrintStackFrame(std::ostream& out, Isolate* isolate,
+                            Local<StackFrame> frame, int index, void* pc);
+static void PrintNativeStack(std::ostream& out);
+#ifndef _WIN32
+static void PrintResourceUsage(std::ostream& out);
+#endif
+static void PrintGCStatistics(std::ostream& out, Isolate* isolate);
+static void PrintSystemInformation(std::ostream& out, Isolate* isolate);
+static void PrintLoadedLibraries(std::ostream& out, Isolate* isolate);
+
+// Global variables
+static int seq = 0;  // sequence number for report filenames
+const char* v8_states[] = {"JS", "GC", "COMPILER", "OTHER", "EXTERNAL", "IDLE"};
+static bool report_active = false;  // recursion protection
+char report_filename[NR_MAXNAME + 1] = "";
+char report_directory[NR_MAXPATH + 1] = "";  // default: current directory
+TIME_TYPE loadtime_tm_struct;  // module load time
+time_t load_time;  // module load time absolute
```

Comment:
For the 5 lines above: This does not look thread-safe

Reply:
will move this to per-isolate

Reply:
subsequent reviews and discussions evolved this into thread-safe as well as well scoped variables.